chainindexer.go

```
@@ -0,0 +1,457 @@
+		}
+
+	}
+	close(workerQuit)
+	a.wg.Done()
+}
+
+// indexCatchUpWorker indexes the transactions of previously validated and
+// stored blocks.
+// NOTE: Must be run as a goroutine
+func (a *addrIndexer) indexCatchUpWorker(workChan chan *indexBlockMsg,
+	wg *sync.WaitGroup, quit chan struct{}) {
+	wg.Add(1)
+out:
+	for {
+		select {
+		case indexJob := <-workChan:
+			addrIndex := a.indexBlockAddrs(indexJob.blk)
+			sha, _ := indexJob.blk.Sha()
+			height := indexJob.blk.Height()
+			a.progressLogger.LogBlockHeight(indexJob.blk)
+			a.writeRequests <- &writeIndexReq{blkHeight: height,
+				blkSha: sha, addrIndex: addrIndex}
+		case <-quit:
+			break out
+		}
+	}
+	wg.Done()
+}
+
+// indexScriptPubKey indexes all data pushes greater than 8 bytes within the
+// passed SPK. Our "address" index is actually a hash160 index, where in the
+// ideal case the data push is either the hash160 of a publicKey (P2PKH) or
+// a Script (P2SH).
+func indexScriptPubKey(addrIndex btcdb.BlockAddrIndex, scriptPubKey []byte,
+	locInBlock *btcwire.TxLoc) error {
+	dataPushes, err := btcscript.PushedData(scriptPubKey)
+	if err != nil {
+		adxrLog.Tracef("Couldn't get pushes: %v", err)
+		return err
+	}
+
+	for _, data := range dataPushes {
+		// Only index pushes greater than 8 bytes.
+		if len(data) < 8 {
+			continue
+		}
+
+		var indexKey [ripemd160.Size]byte
+		// A perfect little hash160.
+		if len(data) <= 20 {
+			copy(indexKey[:], data)
+			// Otherwise, could be a payToPubKey or an OP_RETURN, so we'll
+			// make a hash160 out of it.
+		} else {
+			copy(indexKey[:], btcutil.Hash160(data))
+		}
+
+		addrIndex[indexKey] = append(addrIndex[indexKey], locInBlock)
+	}
+	return nil
+}
+
+// indexBlockAddrs returns a populated index of the all the transactions in the
+// passed block based on the addresses involved in each transaction.
+func (a *addrIndexer) indexBlockAddrs(blk *btcutil.Block) btcdb.BlockAddrIndex {
+	addrIndex := make(btcdb.BlockAddrIndex)
+	txLocs, _ := blk.TxLoc()
```

The error is being ignored here.  That is bad!